### ISSUE TITLE
Debounce updates

### ISSUE THREAD

**Issue by nite** [plotly/dash#395]
Dash would greatly benefit from inbuilt support for debounce. Currently user interaction can easily run into multiple updates or race conditions.

For a simple reproduce, take the datatable example in https://github.com/plotly/dash-table-experiments/blob/master/usage.py, stick some time.sleeps & logging in update handlers, spin it up and select multiple rows in the grid. You should easily see multiple updates. It's tricky in this example to get a race where an earlier user selection comes in last overriding a later selection, however I have seen this in my app. 

Debounce is a standard feature in top javascript util libraries - most js apps feature either lodash or rxjs in their dependencies. This https://www.npmjs.com/package/react-debounce-input could be useful.

ShinyR also has it: https://shiny.rstudio.com/reference/shiny/latest/debounce.html

There has been suggestion of modifying/patching datatable to add debounce, however ideall this should be an inbuilt feature of dash - eg. if we could wrap a whole component, app or section with a dcc.Debounce() control, then that section would debounce all updates from child components.

**Comment by nite** [comment]
If this is more appropriate in https://github.com/plotly/dash-core-components and/or any other updates suggestions for this issue, then I'm happy to move/update it.

**Comment by chriddyp** [comment]
Yeah - I've been writing debounce into some custom components but it would be nice if this was done on the core level so that component authors didn't need to worry about it. I think it could actually live in the `app.callback` signature (which ties into #200 ) and then implemented in `dash-renderer`

**Comment by nite** [comment]
Great - thanks @chriddyp - wish I had some sponsorship powers for you :) happy to collab in other ways though.

**Comment by chriddyp** [comment]
> what custom components have debounce, for reference?

Unfortunately, none that have been open sourced. Only private components that were built as part of custom engagements with customers.

**Comment by nite** [comment]
Could you extract a standalone example component by any chance, or update say datatable with it?
